services/s3/s3.go

```diff
@@ -173,6 +173,123 @@ func (s *S3) HeadObject(input GetObjectInput) (*GetObjectOutput, *awserrors.Erro
 	return s.getObject(input, false)
 }
 
+type ByteRange struct {
+	// inclusive
+	startByte int64
+	// exclusive
+	endByte int64
+}
+
+func parseRangeHeader(rangeHeader string, o *Object) ([]ByteRange, *awserrors.Error) {
+	bytesPrefix := "bytes="
+	if !strings.HasPrefix(rangeHeader, bytesPrefix) {
+		awserrors.XXX_TODO("unknown range header unit")
+	}
+
+	ranges := strings.Split(rangeHeader[len(bytesPrefix):], ",")
+	var result []ByteRange
+
+	for _, encodedRange := range ranges {
+		trimmed := strings.TrimSpace(encodedRange)
+		splitRange := strings.Split(trimmed, "-")
+		if len(splitRange) != 2 {
+			return nil, awserrors.XXX_TODO("invalid range too many -")
+		}
+
+		// indexed from end of resource
+		if splitRange[0] == "" {
+			suffix, err := strconv.ParseInt(splitRange[1], 10, 64)
+			if err != nil {
+				return nil, awserrors.XXX_TODO(err.Error())
+			}
+			result = append(result, ByteRange{
+				startByte: o.ContentLength - suffix,
+				endByte:   o.ContentLength,
+			})
+			continue
+
+		}
+
+		// Suffix
+		if splitRange[1] == "" {
+			startByte, err := strconv.ParseInt(splitRange[0], 10, 64)
+			if err != nil {
+				return nil, awserrors.XXX_TODO(err.Error())
+			}
+			result = append(result, ByteRange{
+				startByte: startByte,
+				endByte:   o.ContentLength,
+			})
+			continue
+		}
+
+		// Both numbers included
+		startByte, err := strconv.ParseInt(splitRange[0], 10, 64)
+		if err != nil {
+			return nil, awserrors.XXX_TODO(err.Error())
+		}
+		endByteInclusive, err := strconv.ParseInt(splitRange[1], 10, 64)
+		if err != nil {
+			return nil, awserrors.XXX_TODO(err.Error())
+		}
+		result = append(result, ByteRange{
+			startByte: startByte,
+			endByte:   endByteInclusive + 1,
+		})
+	}
+	return result, nil
+}
+
+func (s *S3) readerForRange(object *Object, br ByteRange) (io.Reader, *awserrors.Error) {
+	var parts []Part
+	if len(object.Parts) == 0 {
+		parts = []Part{{MD5: object.MD5, Size: object.ContentLength}}
+	} else {
+		parts = object.Parts
+	}
+
+	// bytesUntilStart is zero if we have already passed the start, otherwise it is the total number
+	// of bytes we must skip before starting our response.
+	bytesUntilStart := br.startByte
+	bytesUntilEnd := br.endByte
+	var readers []io.Reader
+	// Loop over parts in order, updating bytesUntilStart and bytesUntilEnd. If the chunk lies
+	// within the range that must be returned, use a section reader to capture the appropriate range.
+	for _, part := range parts {
+		// If we've already passed the end, break
+		if bytesUntilEnd <= 0 {
+			break
+		}
+
+		size := part.Size
+		// If we haven't reached the start yet, skip the chunk.
+		if bytesUntilStart > size {
+			bytesUntilStart -= size
+			bytesUntilEnd -= size
+			continue
+		}
+
+		// Otherwise, open a reader for this chunk.
+		f, err := os.Open(s.filepath(part.MD5))
```

Comment:
preexisting, but we are failing to call Close, so I think this is buggy. I'll fix in a followup